clang/lib/CodeGen/ItaniumCXXABI.cpp

```diff
@@ -3573,32 +3574,8 @@ void ItaniumRTTIBuilder::BuildVTablePointer(const Type *Ty) {
   if (CGM.getItaniumVTableContext().isRelativeLayout())
     VTable = CGM.getModule().getNamedAlias(VTableName);
 
-  // To generate valid device code global pointers should have global address
-  // space in SYCL.
-  bool GenTyInfoGVWithGlobalAS =
-      CGM.getLangOpts().SYCLIsDevice &&
-      CGM.getLangOpts().SYCLAllowVirtualFunctions &&
-      (VTableName == ClassTypeInfo || VTableName == SIClassTypeInfo);
-  auto VTableTy =
-      GenTyInfoGVWithGlobalAS
-          ? CGM.Int8Ty->getPointerTo(
-                CGM.getContext().getTargetAddressSpace(LangAS::sycl_global))
-          : CGM.Int8PtrTy;
-  if (!VTable) {
-    if (GenTyInfoGVWithGlobalAS) {
-      VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy, [&] {
-        return new llvm::GlobalVariable(
-            CGM.getModule(), VTableTy, /*isConstant=*/false,
-            llvm::GlobalVariable::ExternalLinkage, /*Initializer=*/nullptr,
-            VTableName, /*InsertBefore=*/nullptr,
-            llvm::GlobalValue::ThreadLocalMode::NotThreadLocal,
-            llvm::Optional<unsigned>(
-                CGM.getContext().getTargetAddressSpace(LangAS::sycl_global)));
-      });
-    } else {
-      VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy);
-    }
-  }
+  if (!VTable)
+    VTable = CGM.CreateRuntimeVariable(CGM.DefaultInt8PtrTy, VTableName);
```

Comment:
I'm a little confused. Can you explain what this IR means. I know addresspace (1) = global and (4) is generic but what does it mean in here. What is in addresspace (4)?

Reply:
By the way I don't want to hold up this patch. I'm asking all the questions so I understand something I am not very familiar with. Since you have approvals from other FE code owners, you can merge this

Reply:
As I understand it, according to https://llvm.org/docs/LangRef.html#global-variables - here (in my example) addresspace (1) is the address space of the global variable, and (4) is the address space of the pointer it declares. @AlexeySachkov am I wrong?

Reply:
Every global is a pointer in LLVM IR. When you have a pointer as a global variable, you essentially have a pointer to pointer. The top-level address space was incorrectly generated as 0 (private) and that is now fixed. With this patch with have a program-scope variable which is a pointer to generic address space, which in turns contains a pointer to a global address space (that pointer points to actual data somewhere).